reconcile/test/test_ldap_users.py

```diff
@@ -0,0 +1,112 @@
+    )
+    queries_get_users_mock.return_value = [
+        {
+            "org_username": "username1",
+            "requests": [{"path": "test_path1"}],
+            "queries": [{"path": "another_test_path1"}],
+            "gabi_instances": [{"path": "yet_another_test_path1"}],
+            "path": "blah",
+        },
+        {
+            "org_username": "username2",
+            "requests": [{"path": "test_path2"}],
+            "queries": [{"path": "another_test_path2"}],
+            "gabi_instances": [{"path": "yet_another_test_path2"}],
+            "path": "blah",
+        },
+        {
+            "org_username": "username3",
+            "requests": [{"path": "test_path3"}],
+            "queries": [{"path": "another_test_path3"}],
+            "gabi_instances": [{"path": "yet_another_test_path3"}],
+            "path": "blah",
+        },
+    ]
+
+    return queries_get_users_mock
+
+
+@pytest.fixture
+def mocked_queries_get_app_interface_settings(mocker):
+    queries_get_app_interface_settings = mocker.patch.object(
+        ldap_users.queries, "get_app_interface_settings", autospec=True
+    )
+    queries_get_app_interface_settings.return_value = {}
+    return queries_get_app_interface_settings
+
+
+@pytest.fixture
+def mocked_ldap_client(mocker):
+    mock_ldap_client = mocker.patch.object(
+        ldap_users.LdapClient, "from_settings", autospec=True
+    )
+    dummy_ldap_client = mocker.Mock(spec=ldap_users.LdapClient)
+    dummy_ldap_client.__enter__ = dummy_ldap_client
+    dummy_ldap_client.__exit__ = mocker.Mock(return_value=None)
+    dummy_ldap_client.return_value.get_users.return_value = {"username1", "username2"}
+    mock_ldap_client.return_value = dummy_ldap_client
+    return mock_ldap_client
+
+
+@pytest.fixture
+def mocked_mr_client_gateway(mocker):
+    mock_mr_client_gateway = mocker.patch.object(
+        ldap_users.mr_client_gateway, "init", autospec=True
+    )
+    return mock_mr_client_gateway
+
+
+@pytest.fixture
+def mocked_create_delete_user(mocker):
+    mock_create_delete_user = mocker.patch(
+        "reconcile.ldap_users.CreateDeleteUser", autospec=True
+    )
+    return mock_create_delete_user
+
+
+def test_ldap_users_no_dry_run(
+    mocker,
+    mocked_queries_get_users,
+    mocked_queries_get_app_interface_settings,
+    mocked_ldap_client,
+    mocked_mr_client_gateway,
+    mocked_create_delete_user,
+):
+
+    ldap_users.run(False, None)
+
+    mocked_mr_client_gateway.assert_called_once()
+    mocked_create_delete_user.assert_called_once()
+    assert mocked_create_delete_user.call_args == mocker.call(
+        "username3",
+        [
+            {"type": 0, "path": "datablah"},
+            {"type": 1, "path": "datatest_path3"},
+            {"type": 2, "path": "dataanother_test_path3"},
+            {"type": 3, "path": "datayet_another_test_path3"},
+        ],
+    )
+    assert mocked_create_delete_user.method_calls[0][0] == "().submit"
+
+
+def test_ldap_users_dry_run(
```

Comment:
These sorts of tests provide some basic level of testing, but so much has been patched that in reality I'd just be careful about relying on them. To that end, how have you tested this change manually?

Reply:
Yes, I agree. Too much mocking or patching in this case actually can masks certain issues that can rise when the stuff actually runs. I believe an integration test with some fake LDAP server with some user data prepopulated would be better in this case given it's an external dependency and that can involve lot of failure points which aren't usually covered by unit tests in general. 

This is how I went with testing my changes.
- Updated qontract-schema with LDAP schema per [commit](https://github.com/app-sre/qontract-schemas/commit/7329bcf5b4c12cf1fcf49460aed8d7f7f70eb02a)
- Update app-interface to define those settings. You can see the MR in the gitlab project.
- Ran the qontract-server locally followed by a dry run through qontract-server cli. 

Let me know if that makes sense.

Reply:
We could import [ldap-test](https://pypi.org/project/python-ldap-test/) to mock a local LDAP server. I most definitely don't want our tests to try to poke RH's LDAP, but a local server with mock data can help here. Up to you to decide if it's worth the effort.